lib/ecto/adapters/mssql/connection.ex

```diff
@@ -0,0 +1,1644 @@
+          err
+      end
+    end
+
+    def execute(pid, %{} = query, params, opts) do
+      opts = Keyword.put_new(opts, :parameters, params)
+
+      {params, _} =
+        Enum.map_reduce(params, 1, fn param, acc ->
+          {value, type} = prepare_param(param)
+          {%Tds.Parameter{name: "@#{acc}", value: value, type: type}, acc + 1}
+        end)
+
+      opts = Keyword.put(opts, :parameters, params)
+
+      case DBConnection.prepare_execute(pid, query, params, opts) do
+        {:ok, _, query} -> {:ok, query}
+        {:error, _} = err -> err
+      end
+    end
+
+    @impl true
+    def stream(_conn, _sql, _params, _opts) do
+      raise RuntimeError, "Not supported"
+    end
+
+    @impl true
+    def query(conn, sql, params, opts) do
+      {params, _} =
+        Enum.map_reduce(params, 1, fn param, acc ->
+          {value, type} = prepare_param(param)
+          {%Tds.Parameter{name: "@#{acc}", value: value, type: type}, acc + 1}
+        end)
+
+      case Tds.query(conn, sql, params, opts) do
+        {:ok, %Tds.Result{} = result} ->
+          {:ok, Map.from_struct(result)}
+
+        other ->
+          other
+      end
+    end
+
+    @impl true
+    def to_constraints(%Tds.Error{mssql: %{number: code, msg_text: message}}) do
+      Tds.Error.get_constraint_violations(code, message)
+    end
+
+    def to_constraints(_), do: []
+
+    defp prepare_params(params) do
+      params
+      |> Enum.map_reduce(1, fn param, acc ->
+        {value, type} = prepare_param(param)
+        {%Tds.Parameter{name: "@#{acc}", value: value, type: type}, acc + 1}
+      end)
+      |> elem(0)
+    end
+
+    # Boolean
+    defp prepare_param(%Tagged{value: value, type: :boolean}), do: {value, :boolean}
+
+    defp prepare_param(%Tagged{value: value, type: :varchar}), do: {value, :varchar}
+
+    defp prepare_param(%Tagged{value: value, type: :string}), do: {value, :string}
+
+    # UUID and BinaryID
+    defp prepare_param(%Tagged{value: value, type: type}) when type in [:binary_id, :uuid] do
+      case value do
+        <<_::64, ?-, _::32, ?-, _::32, ?-, _::32, ?-, _::96>> ->
+          {:ok, value} = Tds.Types.UUID.dump(value)
+          {value, :uuid}
+
+        any ->
+          {any, :uuid}
+      end
+    end
+
+    defp prepare_param(%Tagged{value: _, tag: Tds.Types.VarChar} = p),
+      do: prepare_param(%{p | type: :varchar})
+
+    defp prepare_param(%Tagged{value: _, tag: Tds.Types.UUID} = p),
+      do: prepare_param(%{p | type: :uuid})
+
+    # any binary
+    defp prepare_param(%Tagged{value: value, type: nil}) when is_binary(value) do
+      type = if String.valid?(value), do: :string, else: :binary
+      {value, type}
+    end
+
+    # Binary
+    defp prepare_param(%Tagged{value: value, type: :binary}), do: {value, :binary}
+
+    # Decimal
+    defp prepare_param(%Decimal{} = value) do
+      {value, :decimal}
+    end
+
+    defp prepare_param(%Tagged{value: {_, {_, _, _, us}} = value, type: :datetime})
```

Comment:
Those datetime types should no longer be used. We can probably remove this and the clause below. :)